File: node/src/providers/gemini-adapter.ts
import type {
  Content,
  GenerateContentRequest,
  GenerateContentResponse,
  ModelParams,
  Part,
  Tool,
} from "@google/generative-ai"

import type {
  Message,
  ProviderUsage,
  StreamEvent,
  TextDelta,
  ToolCall,
  ToolSchema,
  UsageEvent,
} from "../types.js"
import type {
  CanonicalAdapterInput,
  CanonicalMessage,
  CanonicalToolResult,
} from "./content-normalization.js"
import { projectToolOutputToText } from "./content-normalization.js"
import { normalizeToolCall } from "./base.js"
import type { ProtocolRuntimeCapabilities } from "./model-registry.js"
import {
  type AdapterDecodeInput,
  type AdapterOutput,
  type AdapterStreamInput,
  type CanonicalStopReason,
  type ProtocolAdapter,
  ProtocolResponseError,
} from "./protocol-adapter.js"

export const GEMINI_PROTOCOL_CAPABILITIES: ProtocolRuntimeCapabilities = {
  acceptedInputModalities: ["text", "image", "audio"],
  emittedOutputModalities: ["text"],
  tools: true,
  reasoningReplay: "none",
  mediaForms: { imageUrl: true, imageBase64: true, audioBase64: true },
}

export interface GeminiRequestPlan {
  modelParams: ModelParams & Record<string, unknown>
  request: GenerateContentRequest
}

export interface GeminiStreamState {
  readonly input: CanonicalAdapterInput
  readonly toolCalls: Array<{ name: string; args: Record<string, unknown> }>
}

// Google Generate Content streams response chunks, while @google/generative-ai 0.24.1 exposes
// a separate promise for the aggregated response. Candidate finishReason and aggregate usage are
// therefore decoded only by finishStream.
// Source: https://ai.google.dev/api/generate-content#method:-models.streamgeneratecontent
// Source: https://github.com/google-gemini/deprecated-generative-ai-js/blob/v0.24.1/types/responses.ts
function parseArguments(argumentsJson: string): Record<string, unknown> {
  try {
    const value = JSON.parse(argumentsJson) as unknown
    return value && typeof value === "object" && !Array.isArray(value)
      ? value as Record<string, unknown>
      : {}
  } catch {
    return {}
  }
}

function toolName(callId: string, messages: readonly CanonicalMessage[]): string {
  for (let index = messages.length - 1; index >= 0; index--) {
    const match = messages[index].toolCalls?.find(call => call.id === callId)
    if (match) return match.name
  }
  return callId
}

function toolResultPart(
  result: CanonicalToolResult,
  messages: readonly CanonicalMessage[],
): Part {
  return {
    functionResponse: {
      name: toolName(result.callId, messages),
      response: { output: projectToolOutputToText(result.blocks) },
    },
  }
}

function contentPart(block: Exclude<CanonicalMessage["blocks"][number], CanonicalToolResult>): Part | undefined {
  if (block.type === "text") return block.text ? { text: block.text } : undefined
  if (block.type === "image" || block.type === "audio") {
    if (block.source.kind === "base64") {
      return {
        inlineData: {
          mimeType: block.mediaType ?? (block.type === "image" ? "image/png" : "audio/wav"),
          data: block.source.data,
        },
      }
    }
    if (block.source.kind === "url") {
      return {
        fileData: {
          mimeType: block.mediaType ?? (block.type === "image" ? "image/png" : "audio/wav"),
          fileUri: block.source.url,
        },
      }
    }
  }
  if (block.type === "video" || block.type === "file") {
    throw new ProtocolResponseError("gemini", `cannot serialize ${block.type}`)
  }
  throw new ProtocolResponseError("gemini", `cannot serialize ${block.type} source`)
}

export function canonicalGeminiContents(context: CanonicalAdapterInput["context"]): Content[] {
  const messages = context.stateTurn ? [...context.turns, context.stateTurn] : context.turns
  const contents: Content[] = []
  for (const message of messages) {
    const parts: Part[] = []
    for (const call of message.toolCalls ?? []) {
      parts.push({
        functionCall: { name: call.name, args: parseArguments(call.arguments) },
      })
    }
    for (const block of message.blocks) {
      if (block.type === "tool_result") {
        parts.push(toolResultPart(block, messages))
      } else {
        const part = contentPart(block)
        if (part) parts.push(part)
      }
    }
    if (!parts.length) continue
    contents.push({
      role: message.role === "assistant" ? "model" : "user",
      parts,
    })
  }
  return contents
}

function buildTools(tools: readonly ToolSchema[]): Tool[] {
  if (!tools.length) return []
  return [{
    functionDeclarations: tools.map(tool => ({
      name: tool.name,
      description: tool.description,
      parameters: JSON.parse(tool.parameters),
    })),
  }]
}

function modelExtensions(extensions: Readonly<Record<string, unknown>>): Record<string, unknown> {
  const {
    model: _model,
    systemInstruction: _systemInstruction,
    tools: _tools,
    google_search: _googleSearch,
    response_mime_type: _responseMimeType,
    response_schema: _responseSchema,
    generationConfig: _generationConfig,
    ...rest
  } = extensions
  return rest
}

export function geminiVendorConfig(
  extensions: Readonly<Record<string, unknown>>,
): { tools?: Tool[]; generationConfig?: Record<string, unknown> } {
  const tools: Tool[] = []
  if (extensions.google_search) {
    tools.push({
      googleSearch: typeof extensions.google_search === "object"
        ? extensions.google_search as Record<string, unknown>
        : {},
    } as Tool)
  }
  const generationConfig: Record<string, unknown> = {
    ...(extensions.generationConfig as Record<string, unknown> | undefined),
  }
  if (extensions.response_mime_type != null) {
    generationConfig.responseMimeType = extensions.response_mime_type
  }
  if (extensions.response_schema != null) {
    generationConfig.responseSchema = extensions.response_schema
  }
  return {
    ...(tools.length ? { tools } : {}),
    ...(Object.keys(generationConfig).length ? { generationConfig } : {}),
  }
}

function decodeParts(raw: GenerateContentResponse): { content: string; toolCalls: ToolCall[] } {
  const candidate = raw.candidates?.[0]
  let content = ""
  const toolCalls: ToolCall[] = []
  for (const part of candidate?.content.parts ?? []) {
    if (part.text) content += part.text
    else if (part.functionCall) {
      const call = normalizeToolCall(
        part.functionCall.name,
        part.functionCall.name,
        part.functionCall.args,
      )
      if (call) toolCalls.push(call)
    }
  }
  return { content, toolCalls }
}

function numberField(
  raw: Record<string, unknown>,
  field: string,
): number | undefined {
  const value = raw[field]
  if (value === undefined) return undefined
  if (typeof value !== "number" || !Number.isFinite(value) || value < 0) {
    throw new ProtocolResponseError("gemini", `usage.${field} must be a non-negative finite number`)
  }
  return value
}

export class GeminiAdapter implements ProtocolAdapter<
  GeminiRequestPlan,
  GenerateContentResponse,
  GenerateContentResponse,
  GeminiStreamState,
  GenerateContentResponse
> {
  readonly protocol = "gemini" as const
  readonly protocolCapabilities = GEMINI_PROTOCOL_CAPABILITIES

  buildRequest(input: CanonicalAdapterInput): GeminiRequestPlan {
    const extensions = input.extensions
    const vendor = geminiVendorConfig(extensions)
    const tools = [...buildTools(input.tools), ...(vendor.tools ?? [])]
    return {
      modelParams: {
        ...modelExtensions(extensions),
        model: input.resolved.identity.modelId,
        ...(input.context.systemText ? { systemInstruction: input.context.systemText } : {}),
        ...(tools.length ? { tools } : {}),
        ...(vendor.generationConfig ? { generationConfig: vendor.generationConfig } : {}),
      },
      request: { contents: canonicalGeminiContents(input.context) },
    }
  }

  decodeComplete(raw: GenerateContentResponse, _input: AdapterDecodeInput): { message: Message } {
    const decoded = decodeParts(raw)
    const usage = this.normalizeUsage(raw.usageMetadata)
    const rawUsage = raw.usageMetadata as unknown as Record<string, unknown> | undefined
    const tokenCount = usage?.outputTokens
      ?? (rawUsage ? numberField(rawUsage, "totalTokenCount") : undefined)
    return {
      message: {
        role: "assistant",
        content: decoded.content,
        ...(tokenCount !== undefined ? { tokenCount } : {}),
        toolCalls: decoded.toolCalls,
      },
    }
  }

  createStreamState(input: AdapterStreamInput): GeminiStreamState {
    return { input: input.input, toolCalls: [] }
  }

  pushStreamChunk(chunk: GenerateContentResponse, state: GeminiStreamState): AdapterOutput {
    const events: StreamEvent[] = []
    for (const part of chunk.candidates?.[0]?.content.parts ?? []) {
      if (part.text) events.push({ type: "text_delta", delta: part.text } as TextDelta)
      else if (part.functionCall) {
        state.toolCalls.push({
          name: part.functionCall.name,
          args: part.functionCall.args as Record<string, unknown>,
        })
      }
    }
    return { events }
  }

  finishStream(state: GeminiStreamState, final: GenerateContentResponse): AdapterOutput {
    const events: StreamEvent[] = state.toolCalls.map((call, index) => ({
      type: "tool_call",
      id: `call_${index + 1}`,
      name: call.name,
      arguments: call.args,
    }))
    const usage = this.normalizeUsage(final.usageMetadata)
    const rawStopReason = final.candidates?.[0]?.finishReason
    const stopReason = this.normalizeStopReason(rawStopReason)
    if (usage) {
      const raw = final.usageMetadata as unknown as Record<string, unknown>
      const totalTokens = numberField(raw, "totalTokenCount")
        ?? usage.inputTokens + usage.outputTokens
      events.push({
        type: "usage",
        totalTokens,
        inputTokens: usage.inputTokens,
        outputTokens: usage.outputTokens,
        ...(usage.cacheReadInputTokens
          ? { cacheReadInputTokens: usage.cacheReadInputTokens }
          : {}),
        providerUsage: usage,
        ...(stopReason ? { stopReason } : {}),
        ...(rawStopReason ? { rawStopReason } : {}),
      } as UsageEvent)
    }
    return { events }
  }

  normalizeUsage(raw: unknown): ProviderUsage | undefined {
    if (raw === undefined || raw === null) return undefined
    if (typeof raw !== "object" || Array.isArray(raw)) {
      throw new ProtocolResponseError("gemini", "usage must be an object")
    }
    const usage = raw as Record<string, unknown>
    const inputTokens = numberField(usage, "promptTokenCount")
    const outputTokens = numberField(usage, "candidatesTokenCount")
    numberField(usage, "totalTokenCount")
    const cacheReadInputTokens = numberField(usage, "cachedContentTokenCount")
    if (
      inputTokens === undefined
      && outputTokens === undefined
      && cacheReadInputTokens === undefined
    ) return undefined
    return {
      inputTokens: inputTokens ?? 0,
      outputTokens: outputTokens ?? 0,
      ...(cacheReadInputTokens ? { cacheReadInputTokens } : {}),
    }
  }

  normalizeStopReason(raw: string | undefined): CanonicalStopReason | undefined {
    if (raw === undefined) return undefined
    switch (raw) {
      case "STOP":
      case "FINISH_REASON_STOP":
        return "end_turn"
      case "MAX_TOKENS":
        return "max_tokens"
      case "SAFETY":
      case "RECITATION":
      case "BLOCKLIST":
      case "PROHIBITED_CONTENT":
      case "SPII":
        return "content_filter"
      default:
        return "other"
    }
  }
}

File: python/deepstrike/providers/gemini_adapter.py
"""Gemini protocol conversion with a complete request/stream/finalize lifecycle."""
from __future__ import annotations

import json
from dataclasses import dataclass, field
from typing import Any

from deepstrike._kernel import Message, ToolSchema
from deepstrike.providers.base import RenderedContext, UnsupportedModalityError
from deepstrike.providers.protocol_adapter import AdapterOutput, ProtocolResponseError
from deepstrike.providers.stop_reason import canonicalize_stop_reason
from deepstrike.providers.stream import TextDelta, ToolCallEvent, UsageEvent
from deepstrike.providers.usage import ProviderUsage
from deepstrike.types.content import CanonicalAdapterInput, normalize_tool_result, project_tool_output_to_text


def _get(value: Any, name: str) -> Any:
    return value.get(name) if isinstance(value, dict) else getattr(value, name, None)


def _usage_number(raw: Any, field: str) -> int | None:
    value = _get(raw, field)
    if value is None:
        return None
    if isinstance(value, bool) or not isinstance(value, (int, float)) or value < 0:
        raise ProtocolResponseError("gemini", f"usage.{field} must be a non-negative number")
    return int(value)


@dataclass(frozen=True)
class GeminiRequestPlan:
    contents: list[dict]
    config: dict | None


@dataclass
class GeminiStreamState:
    tool_calls: list[dict] = field(default_factory=list)
    last_usage: Any = None
    raw_stop_reason: str | None = None


class GeminiAdapter:
    protocol = "gemini"

    def __init__(self, model: str):
        self._model = model

    def build_contents(self, turns: list[Message]) -> list[dict]:
        contents: list[dict] = []
        for msg in turns:
            if msg.role == "tool":
                parts = []
                for part in getattr(msg, "content_parts", []):
                    if part.type != "tool_result":
                        continue
                    tool_name = part.call_id
                    for turn in reversed(turns):
                        matches = getattr(turn, "tool_calls", None) or []
                        match = next((call for call in matches if call.id == part.call_id), None)
                        if match is not None:
                            tool_name = match.name
                            break
                    parts.append({"function_response": {"name": tool_name, "response": {
                        "output": project_tool_output_to_text(normalize_tool_result(
                            part.call_id, part.output, part.is_error, getattr(part, "content_parts", None),
                        ).blocks),
                    }}})
                if parts:
                    contents.append({"role": "user", "parts": parts})
                continue

            parts: list[dict] = []
            for call in getattr(msg, "tool_calls", None) or []:
                try:
                    args = json.loads(call.arguments)
                except json.JSONDecodeError:
                    args = {}
                parts.append({"function_call": {"name": call.name, "args": args}})
            content_parts = getattr(msg, "content_parts", None) or []
            if content_parts:
                for part in content_parts:
                    if part.type == "text":
                        parts.append({"text": part.text})
                    elif part.type in {"image", "audio"}:
                        media_type = part.media_type or ("image/png" if part.type == "image" else "audio/wav")
                        if getattr(part, "data", None):
                            parts.append({"inline_data": {"mime_type": media_type, "data": part.data}})
                        elif getattr(part, "url", None):
                            parts.append({"file_data": {"mime_type": media_type, "file_uri": part.url}})
                        elif part.type == "audio":
                            raise UnsupportedModalityError("audio", "gemini")
                    elif part.type != "tool_result":
                        raise UnsupportedModalityError(getattr(part, "type", "unknown"), "gemini")
            elif msg.content:
                parts.append({"text": msg.content})
            if parts:
                contents.append({"role": "model" if msg.role == "assistant" else "user", "parts": parts})
        return contents

    @staticmethod
    def build_tools(tools: list[ToolSchema] | tuple[ToolSchema, ...]) -> list[dict] | None:
        if not tools:
            return None
        return [{"function_declarations": [{
            "name": tool.name,
            "description": tool.description,
            "parameters_json_schema": json.loads(tool.parameters),
        } for tool in tools]}]

    def build_config(self, system: str | None, tools: list[ToolSchema] | tuple[ToolSchema, ...], extensions: dict | None) -> dict | None:
        ext = extensions or {}
        config: dict = {}
        if system:
            config["system_instruction"] = system
        tool_defs = list(self.build_tools(tools) or [])
        if ext.get("google_search"):
            tool_defs.append({"google_search": ext["google_search"] if isinstance(ext["google_search"], dict) else {}})
        if tool_defs:
            config["tools"] = tool_defs
            config["automatic_function_calling"] = {"disable": True}
        for key in ("thinking_config", "response_mime_type", "response_schema", "cached_content"):
            if ext.get(key) is not None:
                config[key] = ext[key]
        return config or None

    def build_request(self, input: CanonicalAdapterInput) -> GeminiRequestPlan:
        context = input.context
        turns = [*context.turns, *([context.state_turn] if context.state_turn is not None else [])]
        return GeminiRequestPlan(
            contents=self.build_contents(turns),
            config=self.build_config(context.system_text or None, input.tools, input.extensions),
        )

    @staticmethod
    def _response_parts(response: Any) -> list:
        parts = _get(response, "parts")
        if parts:
            return list(parts)
        candidates = _get(response, "candidates") or []
        content = _get(candidates[0], "content") if candidates else None
        return list(_get(content, "parts") or [])

    @staticmethod
    def _function_call(part: Any) -> Any:
        return _get(part, "function_call")

    def decode_complete(self, raw: Any, input: CanonicalAdapterInput) -> Message:
        content = ""
        tool_calls = []
        for part in self._response_parts(raw):
            text = _get(part, "text")
            if text:
                content += text
                continue
            call = self._function_call(part)
            if call:
                name = str(_get(call, "name") or "")
                args = _get(call, "args") or {}
                if name:
                    from deepstrike.providers.base import normalize_tool_call
                    normalized = normalize_tool_call(name, name, args)
                    if normalized:
                        tool_calls.append(normalized)
        usage = _get(raw, "usage_metadata")
        total = _usage_number(usage, "total_token_count") if usage is not None else None
        return Message(role="assistant", content=content, token_count=total, tool_calls=tool_calls or None)

    def create_stream_state(self, input: CanonicalAdapterInput) -> GeminiStreamState:
        return GeminiStreamState()

    def push_stream_chunk(self, chunk: Any, state: GeminiStreamState) -> AdapterOutput:
        usage = _get(chunk, "usage_metadata")
        if usage is not None:
            state.last_usage = usage
        candidates = _get(chunk, "candidates") or []
        if candidates:
            reason = _get(candidates[0], "finish_reason")
            if isinstance(reason, str) and reason:
                state.raw_stop_reason = reason
        events = []
        for part in self._response_parts(chunk):
            text = _get(part, "text")
            if text:
                events.append(TextDelta(delta=text))
                continue
            call = self._function_call(part)
            if call:
                state.tool_calls.append({"name": str(_get(call, "name") or ""), "args": _get(call, "args") or {}})
        return AdapterOutput(events=events)

    def finish_stream(self, state: GeminiStreamState, final: Any = None) -> AdapterOutput:
        if final is not None:
            self.push_stream_chunk(final, state)
        events = [ToolCallEvent(id=f"call_{index + 1}", name=call["name"], arguments=call["args"])
                  for index, call in enumerate(state.tool_calls)]
        if state.last_usage is None:
            return AdapterOutput(events=events)
        usage = self.normalize_usage(state.last_usage)
        total = _usage_number(state.last_usage, "total_token_count") or 0
        if total:
            events.append(UsageEvent(
                total_tokens=total,
                input_tokens=usage.input_tokens if usage else 0,
                output_tokens=usage.output_tokens if usage else 0,
                cache_read_input_tokens=usage.cache_read_input_tokens if usage else 0,
                stop_reason=canonicalize_stop_reason(state.raw_stop_reason),
                raw_stop_reason=state.raw_stop_reason,
                provider_usage=usage,
            ))
        return AdapterOutput(events=events)

    def normalize_usage(self, raw: Any) -> ProviderUsage | None:
        if raw is None:
            return None
        input_tokens = _usage_number(raw, "prompt_token_count")
        output_tokens = _usage_number(raw, "candidates_token_count")
        cache_read = _usage_number(raw, "cached_content_token_count")
        _usage_number(raw, "total_token_count")
        if input_tokens is None and output_tokens is None and cache_read is None:
            return None
        return ProviderUsage(
            input_tokens=input_tokens or 0,
            output_tokens=output_tokens or 0,
            cache_read_input_tokens=cache_read or 0,
        )
